File: src/modules/graphs/6_topologicalSorting/BFS(kahn's Algorithm).ts
class Graph {
    private vertices: number;
    private adjacencyList: number[][];

    constructor(vertices: number) {
        this.vertices = vertices;
        this.adjacencyList = new Array(vertices).fill(null).map(() => []);
    }

    addEdge(source: number, destination: number): void {
        this.adjacencyList[source].push(destination);
    }

    topologicalSort(): number[] {
        const result: number[] = [];
        const inDegree: number[] = new Array(this.vertices).fill(0);

        // Calculate in-degrees
        for (let i = 0; i < this.vertices; i++) {
            for (const neighbor of this.adjacencyList[i]) {
                inDegree[neighbor]++;
            }
        }

        const queue: number[] = [];
        for (let i = 0; i < this.vertices; i++) {
            if (inDegree[i] === 0) {
                queue.push(i);
            }
        }

        while (queue.length > 0) {
            const vertex = queue.shift() as number;
            result.push(vertex);

            for (const neighbor of this.adjacencyList[vertex]) {
                if (--inDegree[neighbor] === 0) {
                    queue.push(neighbor);
                }
            }
        }

        return result;
    }
}

const vertices: number = 6;
const graph: Graph = new Graph(vertices);

graph.addEdge(5, 2);
graph.addEdge(5, 0);
graph.addEdge(4, 0);
graph.addEdge(4, 1);
graph.addEdge(2, 3);
graph.addEdge(3, 1);

console.log("Topological Sorting:");
const sorted: number[] = graph.topologicalSort();
console.log(sorted.join(" "));

File: src/modules/graphs/6_topologicalSorting/BFS(kahn's Algorithm).go
package main

import "fmt"

type Graph struct {
	vertices      int
	adjacencyList [][]int
}

func newGraph(vertices int) *Graph {
	List := make([][]int, vertices)
	for i := range List {
		List[i] = make([]int, 0)
	}

	return &Graph{
		vertices:      vertices,
		adjacencyList: List,
	}
}

func (g *Graph) AddEdge(source, destination int) {
	g.adjacencyList[source] = append(g.adjacencyList[source], destination)
}

func (g *Graph) topologicalSort() []int {
	result := []int{}
	inDegree := make([]int, g.vertices)

	// calculating indegrees
	for i := 0; i < g.vertices; i++ {
		for _, neighbor := range g.adjacencyList[i] {
			inDegree[neighbor]++
		}
	}

	queue := []int{}
	for i := 0; i < g.vertices; i++ {
		if inDegree[i] == 0 {
			queue = append(queue, i)
		}
	}

	for len(queue) > 0 {
		curr := queue[0]
		queue = queue[1:]

		result = append(result, curr)

		for _, neighbor := range g.adjacencyList[curr] {
			inDegree[neighbor]--
			if inDegree[neighbor] == 0 {
				queue = append(queue, neighbor)
			}
		}
	}

	return result
}

func main() {
	graph := newGraph(6)
	graph.AddEdge(5, 2)
	graph.AddEdge(5, 0)
	graph.AddEdge(4, 0)
	graph.AddEdge(4, 1)
	graph.AddEdge(2, 3)
	graph.AddEdge(3, 1)

	topoSort := graph.topologicalSort()
	fmt.Println(topoSort)
}
